integration/ledger/snapshot_test.go

```diff
@@ -0,0 +1,1049 @@
+			Sequence:        "1",
+			Label:           "marbles",
+		}
+	})
+
+	AfterEach(func() {
+		setup.cleanup()
+		os.RemoveAll(setup.network.RootDir)
+	})
+
+	// Below test does the following when using leveldb as the statedb:
+	// - deploy legacy chaincode
+	// - create snapshots on 2 peers and verify they are same
+	// - bootstrap peers (in existing org and new org) from the snapshot
+	// - verify functions (chaincode invocation/query, channel config update, join new channel, qscc, history)
+	// - upgrade to new lifecycle chaincode
+	// - create a new snapshot from a peer bootstrapped from a snapshot
+	// - bootstrap peers (in existing org and new org) from the new snapshot
+	// - verify functions (chaincode invocation/query, channel config update, join new channel, qscc, history)
+	// - install a new chaincode and verify invocation/query
+	When("levelDB is used as stateDB", func() {
+		It("generates snapshots and bootstraps from snapshot with stateleveldb", func() {
+			org1peer0 := setup.network.Peer("Org1", "peer0")
+			org2peer0 := setup.network.Peer("Org2", "peer0")
+
+			By("deploying legacy chaincode on peers with stateleveldb")
+			nwo.DeployChaincodeLegacy(setup.network, testchannelID, setup.orderer, legacyChaincode)
+
+			height := helper.getLedgerHeight(org1peer0)
+			By(fmt.Sprintf("adding marbles at height %d", height))
+			testKey := "marble-0"
+			channelPeers := setup.network.PeersWithChannel(testchannelID)
+			helper.invokeMarblesChaincode(legacyChaincode.Name, org1peer0, "initMarble", "marble-0", "blue", "35", "tom")
+			nwo.WaitUntilEqualLedgerHeight(setup.network, testchannelID, height+1, channelPeers...)
+			helper.invokeMarblesChaincode(legacyChaincode.Name, org1peer0, "initMarble", "marble-1", "blue", "35", "tom")
+			nwo.WaitUntilEqualLedgerHeight(setup.network, testchannelID, height+2, channelPeers...)
+
+			By("getting txid from a block before snapshot is generated")
+			txid := getTxIDFromLastBlock(setup.network, org1peer0)
+
+			// verify snapshot commands with different parameters")
+			blockNum := nwo.GetLedgerHeight(setup.network, org1peer0, testchannelID) - 1
+			verifySnapshotRequestCmds(setup.network, org1peer0, testchannelID, blockNum)
+
+			// test 1: generate snapshots on 2 peers for the same blockNum and verify they are same
+			_, snapshotDir2 := generateAndCompareSnapshots(setup.network, setup.orderer, org1peer0, org2peer0, legacyChaincode, blockNum)
+
+			// test 2: bootstrap a peer in an existing org
+			By("bootstrapping a peer (org2peer1) in existing org from a snapshot")
+			org2peer1, _ := bootstrapPeerInExistingOrgFromSnapshot(setup, "Org2", "peer1", testchannelID, snapshotDir2, blockNum, false)
+
+			By("installing legacyChaincode on org2peer1")
+			nwo.InstallChaincodeLegacy(setup.network, legacyChaincode, org2peer1)
+
+			availableHistoryResult := verifyPeerAfterJoinBySnapshot(helper, org2peer1, testchannelID, &legacyChaincode, "", blockNum, txid, testKey, []*marbleHistoryResult{})
```

Comment:
- In order to reduce the number of parameters to this function, can you construct a struct (say expectedState) that encapsulates the test data related parameters.

- I guess that it will be better to more readable if you remove the return from this function and always supply the expected history explicitly rather than recycling it - see related comment in this function.